packages/protobuf/src/private/feature-set.ts

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  Edition,
+  FeatureSet,
+  FeatureSetDefaults,
+} from "../google/protobuf/descriptor_pb.js";
+import { protoBase64 } from "../proto-base64.js";
+
+/**
+ * Static edition feature defaults supported by @bufbuild/protobuf.
+ */
+export const featureSetDefaults = FeatureSetDefaults.fromBinary(
+  protoBase64.dec(
+    /*upstream-inject-feature-defaults-start*/ "ChESDAgBEAIYAiABKAEwAhjmBwoREgwIAhABGAEgAigBMAEY5wcKERIMCAEQARgBIAIoATABGOgHIOYHKOgH" /*upstream-inject-feature-defaults-end*/,
+  ),
+);
+
+/**
+ * A merged google.protobuf.FeaturesSet, with all fields guaranteed to be set.
```

Comment:
If a new feature is added, I expect this to happen with the release of a new edition. For an upstream release, we generate the WKT, and also seed the edition defaults (see https://github.com/bufbuild/protobuf-es/pull/619) at build time. The function `createDescriptorSet` uses the seeded defaults. As a result, `createDescriptorSet` can now parse descriptors including the new edition. The message `FeatureSet` will be in sync, since it is part of the WKT.

`createDescriptorSet` may have to be updated to provide a good user experience around a new feature. Then the runtime will have implement this feature. Then we can release full support for the new edition. It is possible to cut a release _before_ we support the new future in the runtime, and only make it available in `createDescriptorSet`, but not in `protoc-gen-es`. We can sort out the exact details if we get there.

If `createDescriptorSet` is called with descriptors using an edition from the future, `createFeatureResolver` will raise an error because the edition is not supported yet. `createDescriptorSet` takes an optional argument to bring your own `FeatureSetDefaults`, allowing to use editions other than the ones supported by the @bufbuild/protobuf release. The `getFeatures()` methods will return `FeatureSet` classes (and still be valid `MergedFeatureSet` because it seems very unlikely that a feature will ever be removed from descriptors). 

The optional argument to bring your own `FeatureSetDefaults` can also be used to provide language-specific extensions to the protobuf message `FeatureSet`. It would be a nice addition to `@bufbuild/protoplugin` to support this use case, but we really ought to support extensions first. 

It's unfortunate that feature resolution is effectively split into build-time and runtime concerns. The logic could certainly be made completely dynamic (and https://github.com/bufbuild/protobuf-es/pull/628 _partially_ gets us there). But since we currently don't support extensions and don't embed descriptors in generated code, implementation is complex and DX is awkward. Since it will also affect bundle size, I think we're better off punting on this now, and reconsider the feature later on.